### ISSUE TITLE
Continuous Deployment was cancelled due to two successive merges

### ISSUE THREAD

**Issue by JordanMartinez** [purescript/purescript#4323]
## Description

See [the recent list of CI builds](https://github.com/purescript/purescript/actions) and [this build job](https://github.com/purescript/purescript/actions/runs/2323384306).

Below is what should have happened.
- #4322 was merged and should have produced `0.15.0-3`.
- Soon thereafter, #4317 was merged and should have produced `0.15.0-4`.
- Sometime much later, #4318 was merged, which should have produced `0.15.0-5`.

However, this is what happened:
- #4322 was merged and started building what would be `0.15.0-3`.
- Soon thereafter, #4317 was merged before the `4322`-build finished
    - this cancelled the `4322`-build
- The `4322+4317` build finished and produced `0.15.0-4`.
- #4318 was merged and started building `0.15.0-4`.
- The `4318` build finished and produced `0.15.0-4`.

We should only short-circuit builds when one pushes a new commit to a pre-existing PR. However, `master` builds should never be cancelled.

**Comment by thomashoneyman** [comment]
A bit of a side note, but should we also only make a new release candidate if files in `src` or `app` were touched?

**Comment by rhendric** [comment]
Why should master builds never be cancelled? Because of the nature of the versioning logic, two master builds can't run in parallel, so our choices are to wait for the previous build to publish, thus delaying the newer prerelease, or to cancel the previous build and get the newer prerelease out sooner, including both changes. I figured the second would be more valuable; do we really need each prerelease to incorporate exactly one new PR?

> A bit of a side note, but should we also only make a new release candidate if files in `src` or `app` were touched?

That's a good idea, though I think we should also consider some additional paths: `npm-package`, `.github/workflows`, `purescript.cabal`, `stack.yaml`?

**Comment by JordanMartinez** [comment]
> Two master builds can't run in parallel, so our choices are to wait for the previous build to publish, thus delaying the newer prerelease, or to cancel the previous build and get the newer prerelease out sooner, including both changes.

Ah... I didn't realize that when making this issue. I'd definitely prefer having a prerelease sooner than one for each commit.

Part of what I was trying to do in this issue is just point out that this situation can occur. So, if a highly desired feature (e.g. source maps) gets almost built before some other PR gets merged, and if that PR breaks something so that one cannot use the built compiler to get the new feature, that would be unfortunate. But... I guess the broken part would be fixed soon thereafter in a new prerelease because using the prerelease would reveal that bug. So, that unfortunate situation is only temporary.

Ok. Nevermind! :sweat_smile: 

> A bit of a side note, but should we also only make a new release candidate if files in src or app were touched?

Agreed.

> though I think we should also consider some additional paths: npm-package, .github/workflows, purescript.cabal, stack.yaml?

So, in total:
- purescript.cabal / stack.yml - because it could affect dependencies, the targets we build, or the flags enabled when building such targets
- src / app - because these directly affect the produced compiler
- npm-package - because this affects the NPM publish
- `.github/workflows` - because these may affect how the binary is built in CI

What about the `bundle` and `ci` dirs? Both of those could affect what CI produces, right? While the `ci` dir may change, I don't think the `bundle` one will.

**Comment by rhendric** [comment]
Looks like #4324 worked; (the build)[https://github.com/purescript/purescript/actions/runs/2328240447] of #4292 did not publish a prerelease. Anything else to do here or can we close this issue?
